numpy/_core/src/multiarray/item_selection.c

```diff
@@ -1599,6 +1600,69 @@ PyArray_Sort(PyArrayObject *op, int axis, NPY_SORTKIND which)
 }
 
 
+/*NUMPY_API
+ * Sort an array in-place with extended parameters
+ */
+NPY_NO_EXPORT int
+PyArray_SortExt(PyArrayObject *op, int axis, bool stable, bool descending, bool nanfirst)
+{
+    PyArray_SortFunc **sort_table = NULL;
+    PyArray_SortFunc *sort = NULL;
+    int n = PyArray_NDIM(op);
+
+    if (check_and_adjust_axis(&axis, n) < 0) {
+        return -1;
+    }
+
+    if (PyArray_FailUnlessWriteable(op, "sort array") < 0) {
+        return -1;
+    }
+
+    sort_table = PyDataType_GetArrFuncs(PyArray_DESCR(op))->sort;
+
+    int which = (nanfirst << 2) + (descending << 1) + stable;
+    // Look for type specific functions
+    switch (which) {
+        case 0:
+            sort = sort_table[0];
+            break;
+        case 1:
+            sort = sort_table[2];
+            break;
+        default:
+            sort = NULL;
+    }
+
+    // Look for appropriate generic function if no type specific version
+    if (sort == NULL) {
+        if (PyDataType_GetArrFuncs(PyArray_DESCR(op))->compare) {
+            switch (which) {
+                case 0:
+                    sort = npy_quicksort;
+                    break;
+                case 1:
+                    sort = npy_timsort;
+                    break;
+                default:
+                    sort = NULL;
+            }
+        }
+        else {
+            PyErr_SetString(PyExc_TypeError,
+                            "type does not have compare function");
+            return -1;
+        }
+    }
+
+    if (sort == NULL) {
+        PyErr_SetString(PyExc_TypeError,
+                        "no current sort function meets the requirements");
+        return -1;
+    }
+    return _new_sortlike(op, axis, sort, NULL, NULL, 0);
+}
```

Comment:
One other question: For the sake of not keeping the old version a stub forever, can we make the old version just call this one?  (Yes, I guess it effectively means deactivating the algorithm choice completely, but I guess that was where we were going anyway.)